[quant][pt2e][xnnpack] XNNPACKQuantizer skip quantization for input and output to workaround histogram observer problem

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113405

Summary:
att, this is because histogram observer does not work for a corner case in mobilebert (observing a scalar tensor of float32 max value)
because histc operator errors out when the value is larger than certain number

Test Plan:
python test/test_quantization.py -k test_mul_float32_max

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant][pt2e][xnnpack] Allow XNNPACKQuantizer to use minmax observer for activation

Summary:
att, this is because histogram observer does not work for a corner case in mobilebert (observing a scalar tensor of float32 max value)

Test Plan:
python test/test_quantization.py -k test_mul_float32_max

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

mcr229 commented on torch/ao/quantization/quantizer/xnnpack_quantizer.py:
```diff
@@ -114,7 +115,7 @@ def get_symmetric_quantization_config(
             act_observer_or_fake_quant_ctr = PlaceholderObserver  # type: ignore[assignment]
         else:
-            act_observer_or_fake_quant_ctr = HistogramObserver  # type: ignore[assignment]
+            act_observer_or_fake_quant_ctr = MinMaxObserver if ptq_act_use_minmax_observer else HistogramObserver  # type: ignore[assignment]
```
Will all observers for mul then be MinMax? It seems like we are just using MinMaxObserver to get around the sys.floatmax, is there a way to use MinMaxObserver only for this case, and use HistogramObserver for the other activations?

to be fair, I don't truly understand how this will affect the numerics

cc. @kimishpatel

jerryzh168 replied:
we can but we'll need to implement https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/quantizer/xnnpack_quantizer.py#L314

this change does not change any existing tests btw, it's an optional flag that allows people to use minmax observer.

for mobilebert in executorch repo, do we have accuracy goals? or is it just to make sure that the quantization and lowering works?

kimishpatel replied:
yeah as I commented in another comment. I dont think this is the right way to do it. Max would you mind fixing this yourself. If HistogramObserver cannot truly handle this and HistogramObsever does not seem the right place to fix this issue, then it belong to quantizer, who should make this decision. @mcr229 if you can do this, that would be great. I dont want this to be a global config.

kimishpatel commented on torch/ao/quantization/quantizer/xnnpack_quantizer.py:
```diff
@@ -114,7 +115,7 @@ def get_symmetric_quantization_config(
             act_observer_or_fake_quant_ctr = PlaceholderObserver  # type: ignore[assignment]
         else:
-            act_observer_or_fake_quant_ctr = HistogramObserver  # type: ignore[assignment]
+            act_observer_or_fake_quant_ctr = MinMaxObserver if ptq_act_use_minmax_observer else HistogramObserver  # type: ignore[assignment]
```
- this doenst seem like the right way to fix this. Also can you explain the underlying issue? If the scalar tensor didnt have float max, but some other value, would that work with histogram observer?

- Orthogonally, why does histogram observer not work with single value or float max?

- For fixing this, I would suggest that we specifically change observer only for nodes/edge, that you are trying to annotate, that has a scalar tensor. And not do that in a blanket manner.

jerryzh168 replied:
* histogram observer is using torch.histc, for scalar tensor, it only works up to some values, not to float_max, the largest that it can work is around float_max/10e15 or something like that
* probably related to histc implementation, I didn't look closely on this
* sure, I think that would require in the user side to just walk through the graph and annotate the nodes before applying global xnnpack quantizer, but this configuration is still needed I think?

kimishpatel replied:
> * histogram observer is using torch.histc, for scalar tensor, it only works up to some values, not to float_max, the largest that it can work is around float_max/10e15 or something like that

Jerry, which op's quantization, via scalar to tensor pass, is triggering this?

jerryzh168 replied:
this is mul/add, both triggers this I think

kimishpatel replied:
> sure, I think that would require in the user side to just walk through the graph and annotate the nodes before applying global xnnpack quantizer, but this configuration is still needed I think?

It shouldnt be done by user. xnnpack quantizer should be able to do this in its annotation logic

kimishpatel reviewed (dismissed):
this needs to happen as per node annotation

jerryzh168 commented:
> this needs to happen as per node annotation

@kimishpatel this is not a change of how the mobilebert is quantized, the main purpose for the PR is to add the configuration, we can do per node annotation when quantizing mobilebert in executorch

jerryzh168 committed:
Update on "[quant][pt2e][xnnpack] Allow XNNPACKQuantizer to use minmax observer for activation"


Summary:
att, this is because histogram observer does not work for a corner case in mobilebert (observing a scalar tensor of float32 max value)

Test Plan:
python test/test_quantization.py -k test_mul_float32_max

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented:
> > this needs to happen as per node annotation
> 
> @kimishpatel this is not a change of how the mobilebert is quantized, the main purpose for the PR is to add the configuration, we can do per node annotation when quantizing mobilebert in executorch

which configuration? `ptq_act_use_minmax_observer` option? I dont think this should be an option for user to configure. It should be done by xnnpack quantizer during annotation:
- Check if the tensor has single element, due to scalar to tensor, and if so extract the value out of the tensor and validate if it is max.
- Alternatively during `transfor_for_annotation` record nodes on which other argument is scalar, and check the value of scalar, and later in annotate, using minmax obserers for those args.

These dont require adding any config

jerryzh168 commented:
> > > this needs to happen as per node annotation
> > 
> > 
> > @kimishpatel this is not a change of how the mobilebert is quantized, the main purpose for the PR is to add the configuration, we can do per node annotation when quantizing mobilebert in executorch
> 
> which configuration? `ptq_act_use_minmax_observer` option? I dont think this should be an option for user to configure. It should be done by xnnpack quantizer during annotation:
> 
> * Check if the tensor has single element, due to scalar to tensor, and if so extract the value out of the tensor and validate if it is max.
> * Alternatively during `transfor_for_annotation` record nodes on which other argument is scalar, and check the value of scalar, and later in annotate, using minmax obserers for those args.
> 
> These dont require adding any configuration

I see, sounds good, I can change annotate_add/annotate_mul etc. directly then, it's not just max that histc fails, as long as it's great than some large value it will fail, maybe we could just use minmax observer for all scalar values? does that sound good?

jerryzh168 committed:
Update on "[quant][pt2e][xnnpack] Allow XNNPACKQuantizer to use minmax observer for activation"


Summary:
att, this is because histogram observer does not work for a corner case in mobilebert (observing a scalar tensor of float32 max value)

Test Plan:
python test/test_quantization.py -k test_mul_float32_max

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e][xnnpack] XNNPACKQuantizer use minmax observer for input and output in some cases"


Summary:
att, this is because histogram observer does not work for a corner case in mobilebert (observing a scalar tensor of float32 max value)
because histc operator errors out when the value is larger than certain number

Test Plan:
python test/test_quantization.py -k test_mul_float32_max

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e][xnnpack] XNNPACKQuantizer use minmax observer for input and output in some cases"


Summary:
att, this is because histogram observer does not work for a corner case in mobilebert (observing a scalar tensor of float32 max value)
because histc operator errors out when the value is larger than certain number

Test Plan:
python test/test_quantization.py -k test_mul_float32_max

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e][xnnpack] XNNPACKQuantizer use FixedQParamsObserver for input and MinMaxObserver for output in some cases"


Summary:
att, this is because histogram observer does not work for a corner case in mobilebert (observing a scalar tensor of float32 max value)
because histc operator errors out when the value is larger than certain number

Test Plan:
python test/test_quantization.py -k test_mul_float32_max

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e][xnnpack] XNNPACKQuantizer use FixedQParamsObserver for input and MinMaxObserver for output in some cases"


Summary:
att, this is because histogram observer does not work for a corner case in mobilebert (observing a scalar tensor of float32 max value)
because histc operator errors out when the value is larger than certain number

Test Plan:
python test/test_quantization.py -k test_mul_float32_max

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

mcr229 reviewed (approved):
Sure I think this looks fine

jerryzh168 committed:
Update on "[quant][pt2e][xnnpack] XNNPACKQuantizer skip quantization for input and output to workaround histogram observer problem"


Summary:
att, this is because histogram observer does not work for a corner case in mobilebert (observing a scalar tensor of float32 max value)
because histc operator errors out when the value is larger than certain number

Test Plan:
python test/test_quantization.py -k test_mul_float32_max

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge
